CollectionNumbersController

JoeCohen opened the pull request:
This PR (when complete) will get CollectionNumbersControllerTest to pass.
It replaces PR #570

JoeCohen commented on app/controllers/collection_numbers_controller.rb:
```diff
@@ -104,6 +104,15 @@ def new
   def create
     store_location
     pass_query_params
+    @layout = calc_layout_params
```
Do we still need `@layout`?
I'm currently finding it only in ImageController; will it go away there?

nimmolo replied:
You're right - Probably not.

JoeCohen commented on app/views/collection_numbers/_form.html.erb:
```diff
@@ -1,29 +1,26 @@
+<%= form_with model: @collection_number do |f| %>
 
-  <%= f.submit button.t, class: "btn mx-auto mt-3" %>
+  <%= submit_tag(local_assigns[:button].t, class: "btn mx-auto mt-3") %>
```
The combination of locals: { button: :SAVE } in the view and button in the form was not working (for reasons I don't understand.

nimmolo replied:
Yes, i've run into that too, but usually with partials-within-partials. Something about the form partial is now called by the new and edit templates, the locals have to be accessed via local_assigns[:attr]. However i would change this to f.submit local_assigns[:button].t, class="btn mx-auto mt-3" so rails can do whatever it wants to do with ID's by binding the submit button to the whole form instance "f"

nimmolo replied:
To give a concrete example i believe this should be
`f.submit local_assigns[:button].t, class: "btn mx-auto mt-3"`
The "f" binds it to the form, and changes the syntax - you don't need the "_tag" at that point

JoeCohen commented on app/views/collection_numbers/_form.html.erb:
```diff
@@ -1,29 +1,29 @@
-<%= form_with model: :collection_number do |f| %>
+<% url =
+     "#{collection_numbers_path}/#{@observation.id}?q=#{get_query_param}" %>
+<%= form_with(model: @collection_number,
```
Is there a more idiomatic way to do this?

nimmolo replied:
Hi Joe- sorry late to the thread! Yes there is. Will supply via email

JoeCohen replied:
You're not late at all. I just started the thread today. It will take me a while to get this ready for review. (Of course, any interim comments are welcome, but you probably have better things to do.)

nimmolo replied:
This shouldn't need a url with the new form_with model syntax, but it will need the @observation.id as a param, i'll post a reference for that syntax in a minute. 

Rails 5.2 just knows that if it's a form_with model: @collection_number, the url is the new_collection_number_path or edit_collection_number_path.

nimmolo replied:
I've actually got the form rewritten with new syntax on my current branch. Here it is. 
The observation.id, necessary for a new collection number, is passed as a hidden field- that's a new thing.

```
<%= form_with model: :collection_number do |f| %>

  <%= f.submit button.t, class: "btn mx-auto mt-3" %>

  <%= f.hidden_field :id, value: @observation.id %>
  <%= f.hidden_field :q, value: get_query_param %>

  <% if @collection_number.observations.size > 1 %>
    <div class="alert alert-warning mt-3">
      <%= :edit_affects_multiple_observations.t(type: :collection_number) %>
    </div>
  <% end %>

  <div class="form-group mt-3">
    <%= f.label :name, :collection_number_name.t + ":") %>
    <span class="help-note">(<%= :required.t %>)</span>
    <%= f.text_field :name, class: "form-control" %>
  </div>

  <div class="form-group mt-3">
    <%= f.label :number, :collection_number_number.t + ":" %>
    <span class="help-note">(<%= :required.t %>)</span>
    <%= f.text_field :number, class:"form-control" %>
  </div>

  <%= f.submit button.t, class: "btn mx-auto mt-3" %>

<% end %>
```

nimmolo replied:
You'll see the new syntax everywhere in this one. All the labels and fields are bound to the |f| instance, and most parentheses are gone.

nimmolo replied:
I'll put a bunch of new controller/view changes up in a few minutes and it may be helpful to merge those into what you're working on. I haven't touched any tests except locations/descriptions, but there is a flotilla of changes throughout views, and some in controllers and helpers. I've gone through them file by file, analphabetically (ok, ok, that's my idea of a good malapropism) and i'm almost back to the top of the alphabet. Will push as soon as that's done.

JoeCohen replied:
FYI: There's a stray close parenthesis on this line:
> ```
>     <%= f.label :name, :collection_number_name.t + ":") %>
> ```

JoeCohen commented on test/controllers/collection_numbers_controller_test.rb:
```diff
@@ -210,14 +206,14 @@ def test_create_collection_number_redirect
-    post(:new, params)
-    assert_redirected_to(obs.show_link_args.merge(q: q))
+    post(:create, params)
+    assert_redirected_to("#{observations_path}/#{obs.id}?q=#{q}")
```
Is there a more idiomatic way to do this?

nimmolo replied:
This should work:
`assert_redirected_to observation_path(obs.id, q: q)`
[Api docs reference](https://api.rubyonrails.org/classes/ActionDispatch/Assertions/ResponseAssertions.html)

nimmolo replied:
Note that the path name is singular for a :show link, plural for :index.
`:show -> observation_path(:id), :index -> observations_path`

nimmolo replied:
This is a good place to use other path syntax, too. You can write (in line 212)
`post new_collection_number_path(params)`
The `new_*` paths can distinguish between the `get` and `post` verbs - 
`get new_collection_number_path` renders the`:new` form, `post new_collection_number_path` `:create`s the collection number, via the same path.
Likewise, for `edit_*` paths,
`get` renders the `:edit` form, `post` `:update`s the collection number, via the same path

JoeCohen replied:
I tried this, but am not getting the syntax write.
Because things aren't actually broken in the current syntax, I'm going to move on and leave this refactoring for another day/PR.

JoeCohen commented:
Thank you. I'll try it out.


El lun., 15 de jun. de 2020 a la(s) 20:04, andrew nimmo (
notifications@github.com) escribió:

> *@nimmolo* commented on this pull request.
> ------------------------------
>
> In test/controllers/collection_numbers_controller_test.rb
> <https://github.com/MushroomObserver/mushroom-observer/pull/571#discussion_r440559037>
> :
>
> >
>      # Prove that post keeps query params intact.
> -    post(:new, params)
> -    assert_redirected_to(obs.show_link_args.merge(q: q))
> +    post(:create, params)
> +    assert_redirected_to("#{observations_path}/#{obs.id}?q=#{q}")
>
> Note that the path name is singular for a :show link, plural for :index.
> :show -> observation_path(:id), :index -> observations_path
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/571#discussion_r440559037>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFGO56OWDDLK57MGZYDRW3OLBANCNFSM4N46KOOQ>
> .
>

JoeCohen commented:
1. When I'm done with a first pass (getting the tests green) I'll pull
this into my branch.
2. I like ending up with more idiomatic and simpler code.
But please consider putting a hold on refactoring until the tests are
green. (cc'ng Nathan and Jason so they can tell me if I'm off base.)

> I definitely recommend focusing on the tests and where possible reducing
> (or at least not expanding) the scope of the PR.

https://groups.google.com/d/msg/mo-developers/fn7V1N1ZsEM/lO3F7OIdAwAJ
<https://groups.google.com/d/msg/mo-developers/fn7V1N1ZsEM/lO3F7OIdAwAJ>

Cf.
https://github.com/MushroomObserver/mushroom-observer/blob/master/README_STANDARDS
:

> = Unit Tests
>
> We are very serious about this nowadays.  Every minute "wasted" writing and
> cleaning up unit tests pays for itself tenfold later.  Nathan is a fan of
> test-driven development.  Even if you don't go that far, no code is
> complete
> until it has a full suite of unit tests verifying its behavior thoroughly.
>
> A full test suite is critical to Agile software development, because it
> gives programmers the confidence to refactor code liberally without
> worrying
> about crashing the website.  If your modifications pass all of our tests,
> you can rest assured that you haven't broken anything.
>
> A related guideline is to force yourself to write a test for every bug you
> find.  Write the test first, ensure that it actually fails, *then* fix it.
>

Maybe:

Does a controller/view combo work?

Get tests green

else

Do the simplest thing possible to make the combo work,

Get tests green

end
Now Refactor


El lun., 15 de jun. de 2020 a la(s) 17:12, andrew nimmo (
notifications@github.com) escribió:

> *@nimmolo* commented on this pull request.
> ------------------------------
>
> In app/views/collection_numbers/_form.html.erb
> <https://github.com/MushroomObserver/mushroom-observer/pull/571#discussion_r440514054>
> :
>
> > @@ -1,29 +1,29 @@
> -<% # form_tag(add_query_param(action: :new, id: @observation.id)) do %>
> -<%= form_with model: :collection_number do |f| %>
> +<% url =
> +     "#{collection_numbers_path}/#{@observation.id}?q=#{get_query_param}" %>
> +<%= form_with(model: @collection_number,
>
> I've actually got the form rewritten with new syntax on my current branch.
> Here it is.
> The observation.id, necessary for a new collection number, is passed as a
> hidden field- that's a new thing.
>
> <%= form_with model: :collection_number do |f| %>
>
>   <%= f.submit button.t, class: "btn mx-auto mt-3" %>
>
>   <%= f.hidden_field :id, value: @observation.id %>
>   <%= f.hidden_field :q, value: get_query_param %>
>
>   <% if @collection_number.observations.size > 1 %>
>     <div class="alert alert-warning mt-3">
>       <%= :edit_affects_multiple_observations.t(type: :collection_number) %>
>     </div>
>   <% end %>
>
>   <div class="form-group mt-3">
>     <%= f.label :name, :collection_number_name.t + ":") %>
>     <span class="help-note">(<%= :required.t %>)</span>
>     <%= f.text_field :name, class: "form-control" %>
>   </div>
>
>   <div class="form-group mt-3">
>     <%= f.label :number, :collection_number_number.t + ":" %>
>     <span class="help-note">(<%= :required.t %>)</span>
>     <%= f.text_field :number, class:"form-control" %>
>   </div>
>
>   <%= f.submit button.t, class: "btn mx-auto mt-3" %>
>
> <% end %>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/571#discussion_r440514054>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFF4QPMQFLYUCJBOXB3RW22HRANCNFSM4N46KOOQ>
> .
>

JoeCohen commented on app/views/collection_numbers/_form.html.erb:
```diff
@@ -1,9 +1,12 @@
-<% # form_tag(add_query_param(action: :new, id: @observation.id)) do %>
-<%= form_with model: :collection_number do |f| %>
+<% id = @observation&.id
```
Something like this is needed to account for the possibility that `@observation` is `nil`. 
I don't know if it happens in real life, but it does happen in collection_numbers_controller_test.

nimmolo replied:
It seems to me like the test should supply that @observation, using `instance_variable_set`. 
If that doesn't work, please let me know, that would be good to know for me too!

JoeCohen commented on app/views/collection_numbers/_form.html.erb:
```diff
@@ -1,9 +1,12 @@
-<%= form_with model: :collection_number do |f| %>
+<% id = @observation&.id
+   id_path_segment = id ? "/#{id}" : nil
+   url = "#{collection_numbers_path}#{id_path_segment}" +
```
Tests expect the form to be submitted to a url that includes `back` and `q` parameters.
Is there a more idiomatic way to make the form do that?

mo-nathan replied:
https://mushroomobserver.org/collection_number/show_collection_number takes me to the collection number index page so it at least seem plausible.  As long as you're tweaking this, I'd rather see this use the correct path methods than hacking together an explicit URL.

nimmolo replied:
Yes - `collection_numbers_path(id)` should produce what you want.

JoeCohen replied:
I don't need to tweak this. The new form syntax handles it fine.

JoeCohen commented on app/views/collection_numbers/_form.html.erb:
```diff
@@ -1,9 +1,12 @@
   <%= f.submit button.t, class: "btn mx-auto mt-3" %>
 
-  <%= f.hidden_field :id, value: @observation.id %>
+  <%= f.hidden_field :id, value: id %>
```
In the new syntax, this line throws an error if `@observation` is `nil`:
```ruby
ERROR["test_edit", #<Minitest::Reporters::Suite:0x000055dd37bb55c0 @name="CollectionNumbersControllerTest">, 11.624770464986796]
 test_edit#CollectionNumbersControllerTest (11.63s)
Minitest::UnexpectedError:         ActionView::Template::Error: no implicit conversion of nil into String
            app/views/collection_numbers/_form.html.erb:3:in `+'
            app/views/collection_numbers/_form.html.erb:3:in `_app_views_collection_numbers__form_html_erb__1435477874758573121_47204341240040'
            app/views/collection_numbers/edit.html.erb:18:in `_app_views_collection_numbers_edit_html_erb__3309629008407258624_47204347943700'
            app/controllers/application_controller.rb:226:in `catch_errors_and_log_request_stats'
            test/functional_test_case.rb:26:in `get'
            test/controller_extensions.rb:113:in `get_with_dump'
            test/controllers/collection_numbers_controller_test.rb:249:in `test_edit'
```

mo-nathan replied:
Would it work to just make adding the hidden field conditional on @observation?  The old code does suggest that this page never actually gets called without @observation being set.

nimmolo replied:
Maybe the test can set `@observation`

JoeCohen replied:
Thank you @mo-nathan & @nimmolo for this discussion and all the other comments above.
1. You've made the penny finally drop for me about how the forms work.
2. The problem in this discussion is fixed in Commit #196f99ca. There were two underlying causes: 
- form failing to set `:id` to `@observation.id`; 
- `edit` action failing to set `@observation`.

JoeCohen commented on app/views/collection_numbers/_form.html.erb:
```diff
@@ -1,9 +1,12 @@
 
-  <%= f.hidden_field :id, value: @observation.id %>
+  <%= f.hidden_field :id, value: id %>
   <%= f.hidden_field :q, value: get_query_param %>
```
Do we all need a hidden field for `:back`?

JoeCohen commented on app/views/collection_numbers/_form.html.erb:
```diff
@@ -1,9 +1,12 @@
 
-  <%= f.hidden_field :id, value: @observation.id %>
+  <%= f.hidden_field :id, value: id %>
   <%= f.hidden_field :q, value: get_query_param %>
```
Do we also need a hidden field for `back`?

nimmolo replied:
That sounds right. I think i did have that at one point... iirc the controller provides it too.

JoeCohen replied:
I don't think we need an explicit hidden `back` field; the following already passes:
```ruby
    # Prove that edit passes "back" and query param through to form.
    get(:edit, params.merge(back: "foo", q: q))
    assert_select("form:match('action', ?)", /\?.*q=#{q}/)
    assert_select("form:match('action', ?)", /\?.*back=foo/)
```

mo-nathan commented on app/controllers/collection_numbers_controller.rb:
```diff
@@ -186,20 +193,17 @@ def show_selected_collection_numbers(query, args = {})
-    if @collection_number.name.blank?
-      flash_error(:create_collection_number_missing_name.t)
+    if @collection_number.name.blank? || @collection_number.number.blank?
+      flash_error(missing_attribute_msg.t)
```
I'd write this something like this:
```
return if missing_attribute
if name_and_number_used?
  ...
```
and then have
```
def missing attribute
  error_message = missing_attribute_error_message
  return false unless error_message
  flash_error(error_message.t)
  return true
end

def missing_attribute_error_message
  return :missing_attribute_error_message if @collection_number.name.blank?
  return :create_collection_number_missing_number if @collection_number.number.blank?
end
```
I haven't tested the above, but I think it's a bit DRYer and more Ruby-ish.

JoeCohen replied:
Done in Commit #b7de553

mo-nathan commented on app/controllers/collection_numbers_controller.rb:
```diff
@@ -262,12 +276,12 @@ def normalize_parameters
       number: @collection_number.number
     ).first
-    !@other_number || @other_number == @collection_number
+    @other_number && @other_number != @collection_number
```
Not a blocker, but I prefer explicit parens over relying on precedence rules when it's not really, really obvious.  In this case I believe that means `@other_number && (@other_number != @collection_number)`

JoeCohen replied:
[Commit #5be922f](https://github.com/MushroomObserver/mushroom-observer/pull/571/commits/5be922f) hopefully finesses/avoids this with:
```ruby
@other_number&. != @collection_number
``` 
I also wonder if
```ruby
@other_number && (@other_number != @collection_number)
```
would be a RuboCop offense, which would mean that we'd to disable and re-enable the cop in order to avoid a complaint and autocorrection.

nimmolo commented:
Hi Joe, I went through and added some further comments. 

Also, I recognize i've introduced conflicts for the merge. Here's what I believe they are: 

I updated the views for collection_numbers last night, you may want to use most of those changes, since they were intended to update the syntax. 

In helpers, I added several variations of your object_path method that i needed for certain situations, particularly object_action_path, which should route to paths for nonstandard CRUD actions in any controller (you'll be able to check those in my version object_link_helper.rb). I believe we'll all find those useful going forward. But they'll need new tests in object_link_helper.rb, which i can supply tonight and merge.

If you prefer, after incorporating Nathan's comments, when you're ready I can take a pass at resolving all the conflicts with my branch.

JoeCohen commented:
Temporarily closing this because merging the current bootstrap4 (Merge pull request #572 from MushroomObserver/nimmo-forms-tests-06062020 #b7cc90ad) breaks some tests.
I will re-open if and when I can get CollectionNumbersControllerTest green.

JoeCohen commented:
> Hi Joe, I went through and added some further comments.
> 
> Also, I recognize i've introduced conflicts for the merge. Here's what I believe they are:
> 
> I updated the views for collection_numbers last night, you may want to use most of those changes, since they were intended to update the syntax.
> 
> In helpers, I added several variations of your `object_path` method that i needed for certain situations, particularly `object_action_path`, which should route to paths for non-CRUD actions in any controller (you'll be able to check those in my version of `object_link_helper.rb`).
> 
> You can use it like this, if you need: `object_action_path(obj, action, params = {})`
> for example, `object_action_path(obj, :show_next, params)`. This example takes advantage of standardizing the naming for `show_next` and `show_prev` across all controllers, but there are other cases where it could be useful. I used it because many actions are shared across Location Descriptions and Name Descriptions, too.
> 
> There are also now `new_object_path` and `edit_object_path` helpers that we can use for new, create, edit, update actions for any unknown object.
> 
> These will need new tests in `test/helpers/object_link_helper.rb`, i need to supply tonight and merge.
> 
> If you prefer, after incorporating Nathan's comments, when you're ready I can take a pass at resolving all the conflicts with my branch.

Nimmo:
Sorry, I just saw this Comment. I already merged and resolved conflicts. 
But I may not have done so correctly.
I'll try to fix the broken tests. In any event I will reopen this tonight.
You should have a close look at how I resolved the conflicts, just in case I broke something or failed to include the changes you mentioned above, particularly the helpers. 
One minor request: If you remember in .rb files, please enclose method arguments in parentheses. See https://docs.rubocop.org/rubocop/cops_style.html#stylemethodcallwithargsparentheses. If you don't remember, don't sweat it because we eventually can get RuboCop to autocorrect this. The main thing is to avoid conflicts (e.g., you do `x_path foo` and I do `x_path(foo)`. This doesn't apply to .erb files, as RuboCop doesn't check them. (And in .erb files, I'm trying to follow your style of omitting the parens.)

JoeCohen commented:
Changed from Closed to Draft. Will change to regular PR tonite at latest, even if there are problems remaining.

JoeCohen commented:
@nimmolo:
This PR is again ready for review.
BUT the branch has one test Error I'm hoping you can fix. (It's beyond me right now.)
```ruby
$ rails t test/controllers/collection_numbers_controller_test.rb -n test_edit_redirect
Started with run options -n test_edit_redirect --seed 20151

ERROR["test_edit_redirect", #<Minitest::Reporters::Suite:0x0000560e74371030 @name="CollectionNumbersControllerTest">, 3.1908181619946845]
 test_edit_redirect#CollectionNumbersControllerTest (3.19s)
Minitest::UnexpectedError:         ActionView::Template::Error: undefined local variable or method `img_proportion' for #<#<Class:0x0000560e74935a98>:0x0000560e7493a2a0>
            app/views/images/_image_thumbnail.html.erb:22:in `_app_views_images__image_thumbnail_html_erb___173300038528238634_47310025935100'
            app/helpers/thumbnail_helper.rb:85:in `thumbnail'
            app/presenters/log_item_presenter.rb:66:in `rss_log_to_presenter'
            app/presenters/log_item_presenter.rb:23:in `initialize'
            app/views/shared/_log_item.html.erb:3:in `new'
            app/views/shared/_log_item.html.erb:3:in `_app_views_shared__log_item_html_erb__2125253696727521290_47310046386980'
            app/views/collection_numbers/edit.html.erb:24:in `block in _app_views_collection_numbers_edit_html_erb___1562936745955419210_47310043170840'
            app/views/collection_numbers/edit.html.erb:23:in `_app_views_collection_numbers_edit_html_erb___1562936745955419210_47310043170840'
            app/controllers/application_controller.rb:226:in `catch_errors_and_log_request_stats'
            test/functional_test_case.rb:26:in `get'
            test/controllers/collection_numbers_controller_test.rb:270:in `test_edit_redirect'
```
(Even if you can't fix it, I suggest merging the PR **after you're otherwise satisfied with it** because it solves a number of problems in this controller and the associated test and views.

JoeCohen commented:
Back to Draft.
I just realized that there are a bunch of review comments I hadn't seen.
I will deal with them in Thurs. a.m., then re-open for review.

nimmolo commented:
Hi Joe, 
Got it. Will check it tomorrow. 
I will also check the img_proportion variable, i added that for UI in the log items.

JoeCohen commented:
> Good morning, Joe
> I believe the thumbnail error might be a `local_assigns[:img_proportion]` thing, but i will need to consult the Rails docs on passing locals. I'd hate to have to use `local_assigns` every time we use a local.
> 
> Does your branch still work if you add the object_link_helpers from my earlier PR?

I'm not getting the syntax right, so I'll leave that for another branch/PR, as this PR passes as-is (other than the thumbnail thing).

I'm going to clean up a few RuboCop offenses, then again request review.

nimmolo commented:
OK - It should use the same syntax. I just added the other methods, and pulled the path definition out to its own method so i could reuse it. But your syntax to call object_path remains the same.

I can work on fixing it after a merge that keeps the new methods. If i merge without the new methods, it will break tests that I just got working in locations/descriptions.

JoeCohen commented:
Again ready for review. (All green except thumbnail thing.)

JoeCohen commented:
Sorry, I don't understand either the context or the potential breakage.
I'm going to take a break (maybe a couple of hours). But I'm probably done
working on this branch.

El jue., 18 de jun. de 2020 a la(s) 09:38, andrew nimmo (
notifications@github.com) escribió:

> OK - It should use the same syntax. I just added the other methods, and
> pulled the path definition out to its own method so i could reuse it. But
> your syntax to call object_path remains the same.
>
> I can work on fixing it after a merge that keeps the new methods. If i
> merge without the new methods, it will break tests that I just got working
> in locations/descriptions.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/571#issuecomment-646152831>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFEQYIGP6MQVBR6CS4TRXI7GVANCNFSM4N46KOOQ>
> .
>

JoeCohen commented:
> Sorry, I don't understand either the context or the potential breakage. I'm going to take a break (maybe a couple of hours). But I'm probably done working on this branch. El jue., 18 de jun. de 2020 a la(s) 09:38, andrew nimmo ( notifications@github.com) escribió:
> […](#)
> OK - It should use the same syntax. I just added the other methods, and pulled the path definition out to its own method so i could reuse it. But your syntax to call object_path remains the same. I can work on fixing it after a merge that keeps the new methods. If i merge without the new methods, it will break tests that I just got working in locations/descriptions. — You are receiving this because you modified the open/close state. Reply to this email directly, view it on GitHub <[#571 (comment)](https://github.com/MushroomObserver/mushroom-observer/pull/571#issuecomment-646152831)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAALDFEQYIGP6MQVBR6CS4TRXI7GVANCNFSM4N46KOOQ> .

PS:  This PR is up-to-date with with bootstrap4, so it includes
```ruby
  def object_path(obj, params = {})
  def edit_object_path(obj, params = {})
  def new_object_path(obj, params = {})
  def object_route_s(obj)
  def object_route_p(obj)
```

nimmolo commented:
Ah - great! Sorry, i didn't see them when i looked, and i should have checked again.

nimmolo commented:
I think `object_link_helpers.rb` still needs one more method, the one that handles non-CRUD actions.
Do you want to add it in your current PR? (will email you about this too.)
```
  def object_action_path(obj, action, params = {})		
     objroute = object_route_p(obj)		
     params[:id] = obj.id		
     send("#{route}_#{action.to_s}_path", params)		
   end
```

JoeCohen commented:
Thanks for the merger! I'll take a break and get back to work in an hour or so.

JoeCohen commented:
> Good morning, Joe
> I believe the thumbnail error might be a `local_assigns[:img_proportion]` thing, but i will need to consult the Rails docs on passing locals. I'd hate to have to use `local_assigns` every time we use a local.
> 
> Does your branch still work if you add the object_link_helpers from my earlier PR?

You're right about the cause being how local variables are passed to partials. 
The identical error was manifested in #587 (currently in Draft), solving it with this change in `thumbnail_helper`, line 85:
>    ~~render "images/image_thumbnail", locals: locals~~
    render "images/image_thumbnail", locals

Thanks for pointing me in the right direction!
